### PR TITLE
Make a status-go nix package

### PR DESCRIPTION
**Pull request by pedropombeiro**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

fixes https://www.pivotaltracker.com/story/show/164984148

### Summary:
[comment]: # (Summarise the problem and how the pull request solves it)

This PR adds a Nix recipe to build status-go (and by extension gomobile, which I expect [to upstream](https://github.com/NixOS/nixpkgs/pull/59018) to nixpkgs). The resulting status-go package contains:

* mobile libraries and headers
* desktop libraries for Linux and macOS
* desktop command line tools (statusd, bootnode, etc)
It is missing the Windows library since that requires a cross-compiler that currently lives in Conan. I hope to make a follow-up PR to either use Conan in Nix to build it in this recipe, or get rid of Conan entirely.

It also updates gradle and uses the Nix package for it, so one less download in the build process.

### Review notes (optional):

I recommend reviewing the PR commit by commit, since they are organized by subject and will make it easier to review.

I've moved the Nix recipes from `/scripts/lib/setup/nix` to `/nix` for ease of use and visibility.

#### Platforms (optional)

* Android
* iOS
* macOS
* Linux
* Windows

#### Areas that maybe impacted (optional)
**Functional**

* 1-1 chats
* wallet / transactions
* dapps / app browsing
* mailservers

status: ready

### CODE REVIEW

**Comment by mandrigin** [comment]
wow, wow, wow!

File: android/app/BUCK
Comment (mandrigin) [inline android/app/BUCK]: that looks like an RN upgrade/downgrade

File: ios/StatusIm/main.m
Comment (mandrigin) [inline ios/StatusIm/main.m]: how are these changes required for status-go dependency?

File: ios/StatusIm/AppDelegate.h
Comment (mandrigin) [inline ios/StatusIm/AppDelegate.h]: how are these changes required for status-go dependency?

**Review by mandrigin — CHANGES_REQUESTED** [review]
I glanced through it and saw some suspicious changes, will need to look closer though

File: ios/StatusIm/main.m
Comment (pedropombeiro) [reply ios/StatusIm/main.m]: They aren't, I just ran a RN eject to check for differences and decided to leave the new header. I'll remove this from the PR.

File: android/app/BUCK
Comment (pedropombeiro) [reply android/app/BUCK]: I'm not sure I understand the comment. I changed all calls to gradlew to instead use Nix's gradle executable.

Comment (mandrigin) [reply android/app/BUCK]: ah, okay, just together with comments that changed, I thought that it was suspicious :)
I didn't link it to status-go, tbh.

File: Makefile
Comment (jakubgs) [inline Makefile]: Why don't we set this somewhere at the top of the `Makefile`? I don't see any reason for it not to be set.

File: modules/react-native-status/android/build.gradle
Comment (jakubgs) [inline modules/react-native-status/android/build.gradle:29]: We remove this stuff because we assume that the `status-go` `aar` file will be under `${System.env.STATUS_GO_LIBDIR}/android`, right?

File: modules/react-native-status/desktop/CMakeLists.txt
Comment (jakubgs) [inline modules/react-native-status/desktop/CMakeLists.txt:49]: This is a weird looking `else`, is this normal?

File: nix/mobile/default.nix
Comment (jakubgs) [inline nix/mobile/default.nix:7]: I don't get where this comes from, this is a string or a variable?

File: scripts/build-desktop.sh
Comment (jakubgs) [inline scripts/build-desktop.sh]: If we are just changing them to variables, shouldn't we just get rid of methods like `getQtBaseBinPathFromNixStore` or `getQtFullOutPathFromNixStore` and use the variables directly?

File: android/gradlew.bat
Comment (jakubgs) [inline android/gradlew.bat:1]: Shouldn't we also remove `android/gradlew`?

**Review by jakubgs — COMMENTED** [review]
I love how much stuff this PR removes.

File: nix/mobile/default.nix
Comment (pedropombeiro) [reply nix/mobile/default.nix:7]: `gradle_4_10` is a variable that comes from nixpkgs: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/gradle/default.nix#L64

Comment (jakubgs) [reply nix/mobile/default.nix:7]: I see, thanks!

File: scripts/build-desktop.sh
Comment (pedropombeiro) [reply scripts/build-desktop.sh]: I was leaving the methods so the name would act as a comment, but I guess we don't really need it. I'll remove them.

File: modules/react-native-status/desktop/CMakeLists.txt
Comment (pedropombeiro) [reply modules/react-native-status/desktop/CMakeLists.txt:49]: It's not how I would choose to do it, but it's how I've seen CMake files being written (both here and at Plex)

Comment (jakubgs) [reply modules/react-native-status/desktop/CMakeLists.txt:49]: Weird as fuck, but what can you do.

File: android/gradlew.bat
Comment (pedropombeiro) [reply android/gradlew.bat:1]: That was my initial approach, but it seems to be used as a marker by RN to check if it has been installed or not, so I just left it. Maybe we could just change the contents to point to Nix's gradle (with a comment), so it's easier to figure out what's happening? Hopefully RN will play along with that.

Comment (jakubgs) [reply android/gradlew.bat:1]: Could work. I never fully got the point of that file. First time I hear it's required by RN.

File: Makefile
Comment (pedropombeiro) [reply Makefile]: Will do that. By the way will remove __toolversion, since that's not used anymore in the Makefile.

File: modules/react-native-status/android/build.gradle
Comment (pedropombeiro) [reply modules/react-native-status/android/build.gradle:29]: right

File: desktop_files/package.json.orig
Comment (mandrigin) [inline desktop_files/package.json.orig]: there's probably no need for this change, and it makes future `git blame` more complicated

File: modules/react-native-status/android/build.gradle
Comment (mandrigin) [inline modules/react-native-status/android/build.gradle]: there is no need for these to be variables anymore, they will be more readable inside `implementation` call later

File: nix/status-go/default.nix
Comment (mandrigin) [inline nix/status-go/default.nix]: why do we print that we are "Building desktop libraries" when we don't? that will be confusing to debug

**Review by mandrigin — APPROVED** [review]
a few smaller things, but that's a great step forward in unification of our buildsystem!

File: desktop_files/package.json.orig
Comment (pedropombeiro) [reply desktop_files/package.json.orig]: I added this because some tool that I ran did this automatically, so we'll need to revert this any time we run that tool. Unfortunately I don't recall which tool that was. I'll revert it.

File: nix/status-go/default.nix
Comment (pedropombeiro) [reply nix/status-go/default.nix]: huh? We *are* building them (just not the Windows ones). I do need to remove the TODO.

Comment (mandrigin) [reply nix/status-go/default.nix]: then this TODO is confusing 😛

**Review by jakubgs — APPROVED** [review]
lqKlotB

**Comment by pedropombeiro** [comment]
Looks like the e2e tests didn't run here?

**Comment by pedropombeiro** [comment]
Added a `STATUS_GO_OWNER` file at the root of the repo which allows devs to customize the fork of status-go to use.

**Comment by churik** [comment]
@PombeirP  - IOS build failed here with error:
```
error: Multiple commands produce '/Users/jenkins/Library/Developer/Xcode/DerivedData/StatusIm-aueaewjajvmwjhfqdvzrrhhxzwwp/Build/Intermediates.noindex/ArchiveIntermediates/StatusIm/IntermediateBuildFilesPath/UninstalledProducts/iphoneos/libyoga.a':
```
https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7918/26/console - can you take a look?

**Comment by pedropombeiro** [comment]
@churik I'm taking a look (hopefully current build fixes it). You can use an earlier build of other platforms for testing, since I'm not changing them with this commit.

**Comment by churik** [comment]
@PombeirP 
Android build also failed with the following error https://ci.status.im/job/status-react/job/prs/job/android/view/change-requests/job/PR-7918/28/ :
```
16:25:08  Testing status-im.test.transport.core
16:25:08  DEBUG [status-im.contact-code.core:63] - initializing contact-code
16:25:08  DEBUG [status-im.contact-code.core:63] - initializing contact-code
16:25:08  DEBUG [status-im.contact-code.core:63] - initializing contact-code
16:25:08  ERROR [status-im.transport.message.core:70] - Something went wrong error 
16:25:08  ERROR [status-im.transport.message.core:70] - Something went wrong nil nil
```
Can you please assist here?

**Comment by pedropombeiro** [comment]
@churik this PR should have no influence on the app behavior or tests, since it's only about building status-go in Nix (and I believe we're building the same version as in develop). Not sure why those tests would be failing tbh.

**Comment by pedropombeiro** [comment]
Started another Android build without timeout, this one should be successful.

**Comment by pedropombeiro** [comment]
> @PombeirP
> Android build also failed with the following error https://ci.status.im/job/status-react/job/prs/job/android/view/change-requests/job/PR-7918/28/ :
> 
> ```
> 16:25:08  Testing status-im.test.transport.core
> 16:25:08  DEBUG [status-im.contact-code.core:63] - initializing contact-code
> 16:25:08  DEBUG [status-im.contact-code.core:63] - initializing contact-code
> 16:25:08  DEBUG [status-im.contact-code.core:63] - initializing contact-code
> 16:25:08  ERROR [status-im.transport.message.core:70] - Something went wrong error 
> 16:25:08  ERROR [status-im.transport.message.core:70] - Something went wrong nil nil
> ```
> 
> Can you please assist here?

That error is also present in develop (nightly): https://ci.status.im/job/status-react/job/combined/job/mobile-android/10690/consoleText

**Comment by churik** [comment]
@PombeirP testing is still in progress, so far:
https://github.com/status-im/status-react/issues/7826 is reproducible again on fresh install.
So I can't start Status app after fresh install on Windows.
Can you take a look?

**Comment by pedropombeiro** [comment]
Looking.

**Comment by pedropombeiro** [comment]
Missing ending backslash in path :-) The next Windows build should be fixed.

**Comment by pedropombeiro** [comment]
OK, all builds passing now after some transient issues in the CI server.

**Comment by churik** [comment]
Thank you @PombeirP! Looking at them now.

**Review by jakubgs — APPROVED** [review]
Can't wait for this to get merged.

**Comment by pedropombeiro** [comment]
Added minor fixup commit to adhere to changes requested for upstream PR in nixpkgs.

**Comment by churik** [comment]
@PombeirP what builds / areas can be affected?

**Comment by pedropombeiro** [comment]
> @PombeirP what builds / areas can be affected?

None, this is just slightly different Nix syntax, but with same output.

**Comment by pedropombeiro** [comment]
I'd say if we want to be completely sure, it's enough to test that status-go is usable on the upcoming iOS build (e.g. we can login)

**Comment by churik** [comment]
Tested on builds with https://github.com/status-im/status-react/commit/bb1397f92f3630648cc70a7aca4cab9205d4d04a (Mac OSx, IOS):
- installing
- restoring account
- log in and join public chat

Tested on builds with https://github.com/status-im/status-react/commit/f668cefde246d6ae6d628566fac0cc042e2a681f (on Windows 10 Home, Linux Ubuntu  18.04, Mac OSx, IOS 11.4, Android 8) randomly:
- fresh install;
- upgrade from latest nightly (desktop);
- chatting in 1-1, public, group chats;
- adding stickers;
- sending transactions from chat and wallet;
- sign transaction for DApp (test DApp);
- adding custom mailserver \ switching to custom mailserver;
- switching off \ on bootnodes, changing networks;
- open links from chat;
- updating profile info;
- PNs;
- sync;
- compatibility (can send \receive messages) with 0.11.0 release.

Did I miss anything @PombeirP ?
If no, I can say that it is ready to merge once `android_e2e` build is completed.

**Comment by pedropombeiro** [comment]
@churik looks pretty complete, thanks!

**Comment by churik** [comment]
@PombeirP https://ci.status.im/blue/organizations/jenkins/status-react%2Fprs%2Fandroid-e2e/detail/PR-7918/39/pipeline/ - stuck?

**Comment by pedropombeiro** [comment]
OK, all builds green, will merge.

**Pull request merged**
